include/bitops.h

```diff
@@ -0,0 +1,224 @@
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#ifndef DOSBOX_BITOPS_H
+#define DOSBOX_BITOPS_H
+
+#include <cassert>
+#include <cstdint>
+#include <limits>
+#include <type_traits>
+
+// An enum and functions to help with bit operations
+// Works with any unisigned integer type up to 32-bits
+// (Extending to 64-bits is currently not needed)
+//
+// Examples
+// ~~~~~~~~
+// uint8_t reg = 0;
+//
+// Setting:
+//   set_bits(reg, bit_0 | bit_1); // 0b0000'0011
+//   set_bits(reg, bit_8); // ASSERT FAIL (out of range)
+//   set_all_bits(reg); // 0b1111'1111
+//
+// Toggling:
+//   toggle_bits(reg, bit_4 | bit_5 | bit_6 | bit_7); // 0b0000'1111
+//   toggle_bits(reg, bit_8); // ASSERT FAIL (out of range)
+//   toggle_all_bits(reg); //  0b1111'0000
+//
+// Clearing:
+//   clear_bits(reg, bit_4 | bit_5); // 0b1100'0000
+//   clear_bits(reg, bit_8); // ASSERT FAIL (out of range)
+//
+// Checking:
+//   are_bits_set(reg, bit_6); // true
+//   are_bits_set(reg, bit_7); // true
+//   are_bits_set(reg, bit_8); // ASSERT FAIL (out of range)
+//   are_bits_set(reg, bit_6 | bit_7); // true (both need to be set)
+//   are_bits_set(reg, bit_5 | bit_6); // false (both aren't set)
+//   are_any_bits_set(reg, bit_5 | bit_6); // true (bit 6 is)
+//   are_bits_cleared(reg, bit_0 | bit_1 | bit_2 | bit_3); // true
+//
+
+enum BITS {
+	bit_0 = 1 << 0,
+	bit_1 = 1 << 1,
+	bit_2 = 1 << 2,
+	bit_3 = 1 << 3,
+	bit_4 = 1 << 4,
+	bit_5 = 1 << 5,
+	bit_6 = 1 << 6,
+	bit_7 = 1 << 7,
+	bit_8 = 1 << 8,
+	bit_9 = 1 << 9,
+	bit_10 = 1 << 10,
+	bit_11 = 1 << 11,
+	bit_12 = 1 << 12,
+	bit_13 = 1 << 13,
+	bit_14 = 1 << 14,
+	bit_15 = 1 << 15,
+	bit_16 = 1 << 16,
+	bit_17 = 1 << 17,
+	bit_18 = 1 << 18,
+	bit_19 = 1 << 19,
+	bit_20 = 1 << 20,
+	bit_21 = 1 << 21,
+	bit_22 = 1 << 22,
+	bit_23 = 1 << 23,
+	bit_24 = 1 << 24,
+	bit_25 = 1 << 25,
+	bit_26 = 1 << 26,
+	bit_27 = 1 << 27,
+	bit_28 = 1 << 28,
+	bit_29 = 1 << 29,
+	bit_30 = 1 << 30,
+	bit_31 = 1 << 31
+};
+
+// Is the bitmask within the maximum value of the register?
+template <typename T>
+static constexpr void validate_bit_width([[maybe_unused]] const T reg, const int bits)
+{
+	// ensure the register is unsigned
+	static_assert(std::is_unsigned<T>::value, "register must be unsigned");
+
+	// Ensure the bits fall within the type size
+	assert(static_cast<uint64_t>(bits) > 0);
+	assert(static_cast<int64_t>(bits) <=
+	       static_cast<int64_t>(std::numeric_limits<T>::max()));
```

Comment:
Would this check break if 64-bit operations are ever allowed?

Reply:
Probably; there might be a better way to catch when the bits argument exceeds the width of the type. That's really all I'm trying to catch. 

( I limited the bits argument to standard 'int' to avoid the 64-bit range, and to avoid compiler warnings when we pass in multiple)

Reply:
Maybe just add a comment to check this if expanding.

Reply:
Added.